src/character.cpp

```diff
@@ -1157,8 +1157,24 @@ bool Character::can_use( const item &it, const item &context ) const
     return true;
 }
 
-void Character::drop_inventory_overflow()
-{
+void Character::drop_invalid_inventory()
+{
+    bool dropped_liquid = false;
+    const_invslice invref = inv.const_slice();
+    for( const std::list<item> *stack : invref ) {
+        const item &it = stack->front();
+        if( it.get_current_phase() == LIQUID ) {
+            const int ipos = get_item_position( &it );
+            i_rem( ipos );
```

Comment:
This needs to be *after* the last usage of `stack` because it removes (and deletes) the object that`stack` points to.
Current accessing `it` below is UB.

Alternatively: `i_rem` returns the removed item, so you can just pass that on to `add_item_or_charges`.

And there is an overload of `i_rem` that takes an item reference and removes the refered item, so you don't need to get the items position yourself, just pass `it` to `i_rem`.

Reply:
I even tried to remind myself not to screw this up after seeing your infinite loop patch.  It's just too easy to do.  I like your filter suggestion below, will go with that.